### PR TITLE
fix(language-server): Fix race conditions when using internal proxy concurrently

### PR DESCRIPTION
**Pull request by PeterSchafer**
## Pull Request Submission

Please check the boxes once done.

The pull request must:

- **Reviewer Documentation**
    - [x] follow [CONTRIBUTING](https://github.com/snyk/cli/blob/main/CONTRIBUTING.md) rules
    - [x] be accompanied by a detailed description of the changes
    - [x] contain a risk assessment of the change (Low | Medium | High) with regards to breaking existing functionality. A change e.g. of an underlying language plugin can completely break the functionality for that language, but appearing as only a version change in the dependencies.
    - [ ] highlight breaking API if applicable
    - [ ] contain a link to the automatic tests that cover the updated functionality.
    - [ ] contain testing instructions in case that the reviewer wants to manual verify as well, to add to the manual testing done by the author.
    - [ ] link to the link to the PR for the User-facing documentation
- **User facing Documentation**
    - [ ] update any relevant documentation in gitbook by submitting a gitbook PR, and including the PR link here
    - [ ] ensure that the message of the final single commit is descriptive and prefixed with either `feat:` or `fix:` , others might be used in rare occasions as well, if there is no need to document the changes in the release notes. The changes or fixes should be described in detail in the commit message for the changelog & release notes.
- **Testing**
    - [ ] Changes, removals and additions to functionality must be covered by acceptance / integration tests or smoke tests - either already existing ones, or new ones, created by the author of the PR.

## Pull Request Review

All pull requests must undergo a thorough review process before being merged. 
The review process of the code PR should include code review, testing, and any necessary feedback or revisions. 
Pull request reviews of functionality developed in other teams only review the given documentation and test reports. 

Manual testing will not be performed by the reviewing team, and is the responsibility of the author of the PR.

For Node projects: It’s important to make sure changes in `package.json` are also affecting `package-lock.json` correctly.

****************************If a dependency is not necessary, don’t add it.**************************** 

When adding a new package as a dependency, make sure that the change is absolutely necessary. We would like to refrain from adding new dependencies when possible.
Documentation PRs in gitbook are reviewed by Snyk's content team. They will also advise on the best phrasing and structuring if needed.

## Pull Request Approval

Once a pull request has been reviewed and all necessary revisions have been made, it is approved for merging into 
the main codebase. The merging of the code PR is performed by the code owners, the merging of the documentation PR 
by our content writers.

## What does this PR do?
When running `snyk language-server` parallel test runs cause race conditions when cleaning up the internal network proxy individually. The current solutions introduces a proxy singleton to avoid concurrent proxy cleanup. 

In addition the `ClearCache()` needed to be extended, since some IDEs might close the application in a way that the normal cleanup at the end of the execution is never reached. To mitigate this issue, the existing cache structure was extended to include a Process ID in the path, which is then used within `ClearCache()` to delete the tmp files that belong to old processes.

For a normal `snyk test` this change causes no functional change, this is why the risk is low.

### CODE REVIEW

File: cliv2/cmd/cliv2/main.go [hunk 1]
```diff
@@ -447,7 +447,9 @@ func displayError(err error, userInterface ui.UserInterface, config configuratio
 			}
 
 			uiError := userInterface.OutputError(err)
-			globalLogger.Err(uiError).Msg("ui failed show error")
+			if uiError != nil {
```
Comment (PeterSchafer) [inline cliv2/cmd/cliv2/main.go:450]: FYI: I just stumbled across this error message which was always written. The change is not related to the actual fix.

File: cliv2/internal/cliv2/cliv2.go [hunk 2]
```diff
@@ -160,6 +163,40 @@ func (c *CLI) ClearCache() error {
...
+				continue
+			}
+
+			p, localError := os.FindProcess(pid)
+
+			if p == nil || localError != nil {
+				continue
+			}
+
+			localError = p.Signal(syscall.Signal(0))
```
Comment (bastiandoetsch) [inline cliv2/internal/cliv2/cliv2.go:190]: What does this do?

Comment (bastiandoetsch) [reply cliv2/internal/cliv2/cliv2.go:190]: Ah got it, you try to contact the pid ;)

File: cliv2/internal/cliv2/cliv2.go [hunk 3]
```diff
@@ -160,6 +163,40 @@ func (c *CLI) ClearCache() error {
...
 
+	// clean up the tmp dir of the current version
+	tempDir := filepath.Dir(c.GetTempDir())
+	fileInfo, err = os.ReadDir(tempDir)
+	if err != nil {
+		return err
+	}
+
+	// cleanup tmp files related to a non-existing process
+	processTempPattern := regexp.MustCompile("([0-9]*)")
```
Comment (bastiandoetsch) [inline cliv2/internal/cliv2/cliv2.go:174]: Is this pattern specific enough?

**Review by bastiandoetsch — APPROVED** [review]

File: cliv2/internal/cliv2/cliv2.go (hunk 3 above)
Comment (PeterSchafer) [reply cliv2/internal/cliv2/cliv2.go:174]: maybe not, I can change this. Thank you!

File: cliv2/internal/cliv2/cliv2.go (hunk 2 above)
Comment (PeterSchafer) [reply cliv2/internal/cliv2/cliv2.go:190]: It is from the golang docs https://pkg.go.dev/os#FindProcess, but the docs are a but thin around this. Looking at the code again, I might need to change this a bit.
